### PR TITLE
gui: Make polling in ClientModel asynchronous

### PR DESCRIPTION
**Pull request by promag**
After #14193 `ClientModel::updateTimer` can take some time, as such the GUI hangs, like #17112.

Fixes this by polling in a background thread and updating the GUI asynchronously.

### CODE REVIEW

**Comment by jonasschnelli** [comment]
Elegant Solution.
Concept ACK.
In the long run, we should probably also improve the 250ms polling on `WalletModel::pollBalanceChanged()`

Will test.

**Comment by promag** [comment]
> also improve the 250ms polling

You mean remove polling? :trollface:

File: src/qt/clientmodel.h [hunk 1]
```diff
@@ -90,7 +90,9 @@ class ClientModel : public QObject
     PeerTableModel *peerTableModel;
     BanTableModel *banTableModel;
 
-    QTimer *pollTimer;
+    //! A thread to interact with m_node asynchronously
+    QThread* const thread;
```
Comment (MarcoFalke) [inline src/qt/clientmodel.h]:
```suggestion
    QThread* const m_thread;
```

:eyes: ?

Comment (promag) [reply src/qt/clientmodel.h]: Sorry, done.

**Comment by laanwj** [comment]
Concept ACK

**Comment by MarcoFalke** [comment]
Put in a gitian build, but I no longer have a windows box for testing

File: src/qt/clientmodel.cpp [hunk 2]
```diff
@@ -90,14 +104,6 @@ int64_t ClientModel::getHeaderTipTime() const
     return cachedBestHeaderTime;
 }
 
-void ClientModel::updateTimer()
-{
-    // no locking required at this point
-    // the following calls will acquire the required lock
-    Q_EMIT mempoolSizeChanged(m_node.getMempoolSize(), m_node.getMempoolDynamicUsage());
-    Q_EMIT bytesChanged(m_node.getTotalBytesRecv(), m_node.getTotalBytesSent());
-}
```
Comment (MarcoFalke) [inline src/qt/clientmodel.cpp]: moving this code into a lambda is stylistic only, right?

**Review by maflcko — COMMENTED** [review]
ACK 8f0d8bd17a575a4235296d82cda80ba268a01dba

Didn't test on windows to confirm this actually reduces the "not responding" pop ups, nor that the mempool size is still updated

File: src/qt/clientmodel.cpp (hunk 2 above)
Comment (promag) [reply src/qt/clientmodel.cpp]: I could keep this here but the connection from the timer needs to be direct (not auto) so that this would run in the timer's slot.

Also, this is now unreachable, and it's a small step towards removing polling.

File: src/qt/clientmodel.cpp [hunk 3]
```diff
@@ -46,6 +56,10 @@ ClientModel::ClientModel(interfaces::Node& node, OptionsModel *_optionsModel, QO
 ClientModel::~ClientModel()
 {
     unsubscribeFromCoreSignals();
+
+    delete pollTimer;
```
Comment (ysangkok) [inline src/qt/clientmodel.cpp]: why is the timer deleted before the thread stops and not the other way around?

Comment (promag) [reply src/qt/clientmodel.cpp]: What is the issue?

Comment (ysangkok) [reply src/qt/clientmodel.cpp]: I don't know if there is an issue or not, I am just wondering if there is any specific reason for this ordering. Mostly because I saw [this comment](https://stackoverflow.com/questions/10492480/starting-qtimer-in-a-qthread#comment73988017_10493492).

Comment (promag) [reply src/qt/clientmodel.cpp]: Thanks for pointing it, there's indeed an issue here - must be deleteLater of move after joining the other thread.

Comment (promag) [reply src/qt/clientmodel.cpp]: Fixed.

**Comment by promag** [comment]
@MarcoFalke @Sjors force push with
```diff
--- a/src/qt/clientmodel.cpp
+++ b/src/qt/clientmodel.cpp
@@ -57,9 +57,9 @@ ClientModel::~ClientModel()
 {
     unsubscribeFromCoreSignals();

-    delete pollTimer;
     m_thread->quit();
     m_thread->wait();
+    delete pollTimer;
 }

 int ClientModel::getNumConnections(unsigned int flags) const
```

**Comment by luke-jr** [comment]
"Moved to another thread" isn't really "asynchronous", but Concept ACK

File: src/qt/clientmodel.cpp [hunk 4]
```diff
@@ -30,14 +31,23 @@ ClientModel::ClientModel(interfaces::Node& node, OptionsModel *_optionsModel, QO
...
     peerTableModel = new PeerTableModel(m_node, this);
     banTableModel = new BanTableModel(m_node, this);
-    pollTimer = new QTimer(this);
-    connect(pollTimer, &QTimer::timeout, this, &ClientModel::updateTimer);
+
+    pollTimer = new QTimer;
+    // move timer to thread so that polling doesn't disturb event loop
+    pollTimer->moveToThread(m_thread);
+    m_thread->start();
+    connect(pollTimer, &QTimer::timeout, [this] {
```
Comment (luke-jr) [inline src/qt/clientmodel.cpp]: Can't you leave `ClientModel::updateTimer` as-is for now? Or at least move the code movement to another commit?

File: src/qt/clientmodel.cpp [hunk 5]
```diff
@@ -30,14 +31,23 @@ ClientModel::ClientModel(interfaces::Node& node, OptionsModel *_optionsModel, QO
...
+    m_thread(new QThread(this))
 {
     cachedBestHeaderHeight = -1;
     cachedBestHeaderTime = -1;
     peerTableModel = new PeerTableModel(m_node, this);
     banTableModel = new BanTableModel(m_node, this);
-    pollTimer = new QTimer(this);
-    connect(pollTimer, &QTimer::timeout, this, &ClientModel::updateTimer);
+
+    pollTimer = new QTimer;
```
Comment (luke-jr) [inline src/qt/clientmodel.cpp]: Why can't this use `QTimer(m_thread)`?

File: src/qt/clientmodel.cpp (hunk 4 above)
Comment (promag) [reply src/qt/clientmodel.cpp]: Yes but I'd have to connect with Qt::Direct connection, otherwise the slot would be called in the GUI event loop.

**Comment by promag** [comment]
@luke-jr right, conceptually asynchronous from the GUI thread.

File: src/qt/clientmodel.cpp (hunk 5 above)
Comment (promag) [reply src/qt/clientmodel.cpp]: Because Qt threading model :trollface:

Being a child of QThread does nothing other than be owned by it. To have it's events processed in the thread then moveToThread must be used. This also works only if the object being moved has no parent.

Comment (ysangkok) [reply src/qt/clientmodel.cpp]: why can't this be a unique_ptr?

Comment (luke-jr) [reply src/qt/clientmodel.cpp]: I thought it deleted children when being deleted itself?

File: src/qt/clientmodel.cpp (hunk 4 above)
Comment (luke-jr) [reply src/qt/clientmodel.cpp]: Even after `moveToThread`? Because of the slot being on a GUI object?

Comment (promag) [reply src/qt/clientmodel.cpp]: Yes, with auto connection (the default) the signal checks if the target's thread is the same as the sender or not, if not then it's the same as queued connection.

File: src/qt/clientmodel.cpp (hunk 5 above)
Comment (promag) [reply src/qt/clientmodel.cpp]: You can't have a tree of objects with different thread affinity.

Comment (promag) [reply src/qt/clientmodel.cpp]: > why can't this be a unique_ptr?

It can, but it requires more changes, which are unnecessary in the scope of this PR.

**Comment by fjahr** [comment]
tested ACK 8f0d8bd 

Built, ran tests, some manual testing while running for 24h on macOS.

Please add a description. I got the context from the [core dev logs](http://www.erisian.com.au/bitcoin-core-dev/log-2019-10-17.html) but there should be some information here as well.

File: src/qt/clientmodel.cpp [hunk 6]
```diff
@@ -46,6 +56,10 @@ ClientModel::ClientModel(interfaces::Node& node, OptionsModel *_optionsModel, QO
 ClientModel::~ClientModel()
 {
     unsubscribeFromCoreSignals();
+
+    m_thread->quit();
+    m_thread->wait();
+    delete pollTimer;
```
Comment (laanwj) [inline src/qt/clientmodel.cpp]: As discussed on IRC: I don't think deleting the timer in the main thread, after having moved it to another thread is safe. Qt objects need to be deleted in the thread that owns them. I think that that thread doesn't exist anymore here doesn't change that. You're not even supposed to hold on to the pointer.

I think you can do this before moving it to the thread to make sure it's cleaned up:
```
connect(&thread, &QThread::finished, pollTimer, &QTimer::deleteLater);
```
(like in `RPCConsole::startExecutor()`)

File: src/qt/clientmodel.cpp (hunk 4 above)
Comment (laanwj) [reply src/qt/clientmodel.cpp]: I think you should do this connect before moving the timer to the thread?

Comment (promag) [reply src/qt/clientmodel.cpp]: `QObject::connect` is thread safe.

File: src/qt/clientmodel.cpp (hunk 6 above)
Comment (promag) [reply src/qt/clientmodel.cpp]: Done.

**Review by promag — COMMENTED** [review]
Updated after IRC discussion.

File: src/qt/clientmodel.cpp [hunk 7]
```diff
@@ -30,22 +31,36 @@ ClientModel::ClientModel(interfaces::Node& node, OptionsModel *_optionsModel, QO
...
+    QTimer* timer = new QTimer;
+    timer->setInterval(MODEL_UPDATE_DELAY);
+    connect(timer, &QTimer::timeout, [this] {
+        // no locking required at this point
+        // the following calls will acquire the required lock
+        Q_EMIT mempoolSizeChanged(m_node.getMempoolSize(), m_node.getMempoolDynamicUsage());
+        Q_EMIT bytesChanged(m_node.getTotalBytesRecv(), m_node.getTotalBytesSent());
+    });
+    connect(m_thread, &QThread::finished, timer, &QObject::deleteLater);
+    connect(m_thread, &QThread::started, timer, static_cast<void (QTimer::*)()>(&QTimer::start));
```
Comment (promag) [inline src/qt/clientmodel.cpp]: The cast is needed because `QTimer::start` is overloaded.

**Review by Sjors — APPROVED** [review]
ACK 092d91f6196af497b9618418f84527f552ef1154. Tested without a wallet (`bitcoin-qt -nowallet`) by calling `invalidateblock` and later `reconsiderblock`. Without this commit QT is very unresponsive during sync. The original issue seems less pronounced with a wallet.

Would it make sense to create a NodePoller object, which handles the timer life cycle? Or is that overkill if the plan is to get rid of polling asap anyway?

**Comment by laanwj** [comment]
> Would it make sense to create a NodePoller object, which handles the timer life cycle? Or is that overkill if the plan is to get rid of polling asap anyway?

I think that as long as polling is necessary (and I don't believe we will completely get rid of it any time soon) it would make sense to have one thread to do all polling, and not, say, create a thread per WalletModel too.

But it's out of scope for this PR.

File: src/qt/clientmodel.cpp (hunk 7 above)
Comment (laanwj) [reply src/qt/clientmodel.cpp]: could this work instead of the scary cast ?
```c++
connect(m_thread, &QThread::started, timer, [timer] {
    timer->start();
}
```

Comment (promag) [reply src/qt/clientmodel.cpp]: Yes that works, and the following also works:
```cpp
connect(m_thread, &QThread::started, [timer] {
    timer->start();
});
```
because `started` is emitted from the associated thread.

Comment (laanwj) [reply src/qt/clientmodel.cpp]: Much better!

File: src/qt/clientmodel.cpp [hunk 8]
```diff
@@ -30,22 +31,36 @@ ClientModel::ClientModel(interfaces::Node& node, OptionsModel *_optionsModel, QO
...
+
+    QTimer* timer = new QTimer;
+    timer->setInterval(MODEL_UPDATE_DELAY);
+    connect(timer, &QTimer::timeout, [this] {
+        // no locking required at this point
+        // the following calls will acquire the required lock
+        Q_EMIT mempoolSizeChanged(m_node.getMempoolSize(), m_node.getMempoolDynamicUsage());
+        Q_EMIT bytesChanged(m_node.getTotalBytesRecv(), m_node.getTotalBytesSent());
+    });
+    connect(m_thread, &QThread::finished, timer, &QObject::deleteLater);
```
Comment (promag) [inline src/qt/clientmodel.cpp]: Ops, I think this is wrong - `deleteLater` isn't called because the thread's event loop is not running. I'll check this.

Comment (promag) [reply src/qt/clientmodel.cpp]: Nope, all good! From https://doc.qt.io/qt-5/qthread.html#finished:
> When this signal is emitted, the event loop has already stopped running. No more events will be processed in the thread, except for deferred deletion events. This signal can be connected to QObject::deleteLater(), to free objects in that thread.

Comment (laanwj) [reply src/qt/clientmodel.cpp]: Phew, good to know

**Comment by laanwj** [comment]
ACK 6b6be41c36e4fe9a74bed50e7f0a06532ab1260b

**Comment by Sjors** [comment]
Code review re-ACK 6b6be41; only replaced the scary cast with `{ timer->start(); }`
